utils/sample_qc.py

```diff
@@ -0,0 +1,487 @@
+
+
+
+def compute_stratified_metrics_filter(ht: hl.Table, qc_metrics: List[str], strata: List[str] = None) -> hl.Table:
+    """
+    Compute median, MAD, and upper and lower thresholds for each metric used in pop- and platform-specific outlier filtering
+
+    :param MatrixTable ht: HT containing relevant sample QC metric annotations
+    :param list qc_metrics: list of metrics for which to compute the critical values for filtering outliers
+    :param list of str strata: List of annotations used for stratification. These metrics should be discrete types!
+    :return: Table grouped by pop and platform, with upper and lower threshold values computed for each sample QC metric
+    :rtype: Table
+    """
+
+    def make_pop_filters_expr(ht: hl.Table, qc_metrics: List[str]) -> hl.expr.SetExpression:
+        return hl.set(hl.filter(lambda x: hl.is_defined(x),
+                                [hl.or_missing(ht[f'fail_{metric}'], metric) for metric in qc_metrics]))
+
+    ht = ht.select(*strata, **ht.sample_qc.select(*qc_metrics)).key_by('s').persist()
+
+    def get_metric_expr(ht, metric):
+        return hl.bind(
+            lambda x: x.annotate(
+                upper=x.median + 4 * x.mad,
+                lower=x.median - 4 * x.mad
+            ),
+            hl.bind(
+                lambda elements, median: hl.struct(
+                    median=median,
+                    mad=1.4826 * hl.median(hl.abs(elements - median))
+                ),
+                *hl.bind(
+                    lambda x: hl.tuple([x, hl.median(x)]),
+                    hl.agg.collect(ht[metric])
+                )
+            )
+        )
+
+    agg_expr = hl.struct(**{metric: get_metric_expr(ht, metric) for metric in qc_metrics})
+    if strata:
+        ht = ht.annotate_globals(metrics_stats=ht.aggregate(hl.agg.group_by(hl.tuple([ht[x] for x in strata]), agg_expr)))
+    else:
+        ht = ht.annotate_globals(metrics_stats={(): ht.aggregate(agg_expr)})
+
+    strata_exp = hl.tuple([ht[x] for x in strata]) if strata else hl.tuple([])
+
+    fail_exprs = {
+        f'fail_{metric}':
+            (ht[metric] >= ht.metrics_stats[strata_exp][metric].upper) |
+            (ht[metric] <= ht.metrics_stats[strata_exp][metric].lower)
+        for metric in qc_metrics}
+    ht = ht.transmute(**fail_exprs)
+    pop_platform_filters = make_pop_filters_expr(ht, qc_metrics)
+    return ht.annotate(pop_platform_filters=pop_platform_filters)
+
+
+def flatten_duplicate_samples_ht(dups_ht: hl.Table) -> hl.Table:
+    """
+    Flattens the result of `filter_duplicate_samples`, so that each line contains a single sample.
+    An additional annotation is added: `dup_filtered` indicating which of the duplicated samples was kept.
+
+    :param Table dups_ht: Input HT
+    :return: Flattened HT
+    :rtype: Table
+    """
+    dups_ht = dups_ht.key_by()
+    dups_ht = dups_ht.annotate(
+        dups=hl.array([(dups_ht.s_kept, True)]).extend(
```

Comment:
This is probably fine, but I was a little unsure so I figured I would ask. Will the table returned by `filter_duplicate_samples` always include the s_kept?

Reply:
Good  point -- no this could  vary!